napari/components/viewer_model.py

```diff
@@ -995,9 +995,16 @@ def _add_layers_with_plugins(
             )
             # actually add the layer
             new = self._add_layer_from_data(*_data)
+
             # some add_* methods return a List[Layer], others just a Layer
             # we want to always return a list
-            added.extend(new if isinstance(new, list) else [new])
+            new = new if isinstance(new, list) else [new]
+
+            # store the source of the layer
+            for layer in new:
+                layer.source = filename
```

Comment:
I think we should add `layer.source = None` to the [base layer](https://github.com/napari/napari/blob/v0.4.2/napari/layers/base/base.py) with documentation in the attributes is always present.

Also curious how people would feel about making it a read only property with no setter and then using `layer._source = filename` internally here.

Inside the base class it would look like something like

```python
    @property
    def source(self) -> Optional[str]:
        """Return source of data if loaded from a file. None otherwise.
        """
        return self._source
```

Reply:
I'm for this.  Read only good too for now.  More complex usage can come later as use cases arises